src/parsers/psse_dynamic_data.jl

```diff
@@ -0,0 +1,267 @@
+
+Each dictionary indexed by id contains a vector with 5 of its components:
+* Machine
+* Shaft
+* AVR
+* TurbineGov
+* PSS
+
+Files must be parsed from a .raw file (PTI data format) and a .dyr file.
+
+## Examples:
+```julia
+raw_file = "Example.raw"
+dyr_file = "Example.dyr"
+sys = System(raw_file, dyr_file)
+```
+
+"""
+function System(sys_file::AbstractString, dyr_file::AbstractString; kwargs...)
+    ext = splitext(sys_file)[2]
+    if lowercase(ext) in [".raw"]
+        pm_kwargs = Dict(k => v for (k, v) in kwargs if !in(k, SYSTEM_KWARGS))
+        sys = System(PowerModelsData(sys_file; pm_kwargs...); kwargs...)
+    else
+        throw(DataFormatError("$sys_file is not a .raw file type"))
+    end
+    bus_dict_gen = _parse_dyr_components(dyr_file)
+    add_dyn_injectors!(sys, bus_dict_gen)
+    return sys
+end
+
+"""
+Add to a system already created the dynamic components. 
+The system should already be parsed from a .raw file.
+
+## Examples:
+```julia
+dyr_file = "Example.dyr"
+add_dyn_injectors!(sys, dyr_file)
+```
+
+"""
+function add_dyn_injectors!(sys::System, dyr_file::AbstractString)
+    bus_dict_gen = _parse_dyr_components(dyr_file)
+    add_dyn_injectors!(sys, bus_dict_gen)
+end
+
+function add_dyn_injectors!(sys::System, bus_dict_gen::Dict)
+    @info "Generators provided in .dyr, without a generator in .raw file will be skipped."
+    for g in get_components(ThermalStandard, sys)
+        _num = get_number(get_bus(g))
+        _name = get_name(g)
+        _id = parse(Int64, split(_name, "-")[end])
+        temp_dict = get(bus_dict_gen, _num, nothing)
+        if isnothing(temp_dict)
+            @warn "Generator at bus $(_num), id $(_id), not found in Dynamic Data.\nVoltage Source will be used to model it."
+            r, x = get_ext(g)["z_source"]
+            if x == 0.0
+                @warn "No series reactance found. Setting it to 1e-6"
+                x = 1e-6
+            end
+            s = _make_source(g, r, x)
+            #Remove ThermalGenerator
+            remove_component!(typeof(g), sys, _name)
+            #Add Source
+            add_component!(sys, s)
+        else
+            #Obtain Machine from Dictionary
+            machine = temp_dict[_id][1]
+            #Update R,X from RSORCE and XSORCE from RAW file
+            r, x = get_ext(g)["z_source"]
+            set_R!(machine, r)
+            #Obtain Shaft from dictionary
+            shaft = temp_dict[_id][2]
+            if typeof(machine) == BaseMachine
+                if x == 0.0
+                    @warn "No series reactance found. Setting it to 1e-6"
+                    x = 1e-6
+                end
+                set_Xd_p!(machine, x)
+                if get_H(shaft) == 0.0
+                    @info "Machine at bus $(_num), id $(_id) has zero inertia. Modeling it as Voltage Source"
+                    s = _make_source(g, r, x)
+                    #Remove ThermalGenerator
+                    remove_component!(typeof(g), sys, _name)
```

Comment:
You could call `remove_component!(sys, g)` instead.  Is this safe in Julia?  You are removing an object from a dictionary while iterating over that dictionary.  Calling `collect(get_components())` would guarantee safety.  Or build a vector of generators to remove.